compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs

```diff
@@ -147,45 +151,81 @@ createProjectPackageDb projectRoot (disableScenarioService -> opts) modulePrefix
-              workDir
-              dbPath
-              projectPackageDatabase
-              pkgId
-              pkgName
-              mbPkgVersion
-              deps
-              dependenciesInPkgDb
-              exposedModules
+        -- unless (pkgId `Set.member` stablePkgIds || pkgId `Set.member` dependenciesInPkgDbIds) $ do
+        case () of
+          _ | pkgId `Set.member` stablePkgIds -> pure ()
+          _ | pkgId `Set.member` baseDependenciesIds -> pure ()
+          _ | pkgId `Set.member` depPkgIds -> do
+            registerDepInPkgDb (dalf pkgNode) depsDir dbPath
+          _ | pkgId `Set.member` dataDepPkgIds -> do
+            PackageMap dependenciesSoFar <-
+              (>>= maybe (fail "Failed to generate package info") pure) $
+                withDamlIdeState opts loggerH diagnosticsLogger $ \ide -> runActionSync ide $ runMaybeT $
+                  fst <$> useE GeneratePackageMap projectRoot
+
+            exposedModules <- getExposedModules opts projectRoot
+
+            let unitIdStr = unitIdString $ unitId pkgNode
+            let pkgIdStr = T.unpack $ LF.unPackageId pkgId
+            let (pkgName, mbPkgVersion) = LF.splitUnitId (unitId pkgNode)
+            let deps =
+                    [ (unitId depPkgNode, dalfPackage depPkgNode)
+                    | (depPkgNode, depPkgId) <- map vertexToNode $ reachable depGraph vertex
+                    , pkgId /= depPkgId
+                    , not (depPkgId `Set.member` stablePkgIds)
+                    ]
+            let workDir = dbPath </> unitIdStr <> "-" <> pkgIdStr
+            let dalfDir = dbPath </> pkgIdStr
+            createDirectoryIfMissing True workDir
+            createDirectoryIfMissing True dalfDir
+            BS.writeFile (dalfDir </> unitIdStr <.> "dalf") $ LF.dalfPackageBytes $ dalfPackage pkgNode
+
+            generateAndInstallIfaceFiles
+                (LF.extPackagePkg $ LF.dalfPackagePkg $ dalfPackage pkgNode)
+                (stubSources pkgNode)
+                opts
+                workDir
+                dbPath
+                projectPackageDatabase
+                pkgId
+                pkgName
+                mbPkgVersion
+                deps
+                dependenciesSoFar
+                exposedModules
+          _ -> do
+            putStrLn $ "default: " <> show (pkgId, unitId pkgNode)
+
+      PackageMap dependenciesEnd <-
+        (>>= maybe (fail "Failed to generate package info") pure) $
+          withDamlIdeState opts loggerH diagnosticsLogger $ \ide -> runActionSync ide $ runMaybeT $
+            fst <$> useE GeneratePackageMap projectRoot
+
+      let dependencyInfo = DependencyInfo
+            { dependenciesInPkgDb = dependenciesEnd
+            , dalfsFromDependencies
+            , dalfsFromDataDependencies
+            , mainUnitIds
+            }
+
+      -- We perform these check before checking for unit id collisions
+      -- since it provides a more useful error message.
+      whenLeft
```

Comment:
since these checks depend on `dependencyInfo`, which is only correct after the last dependency has been registered, they need to happen at the end. I don't like that since it means the compiler might do a lot of work only to fail at the end :/